### PR TITLE
Allow leading zeros in numeric literals

### PR DESCRIPTION
**Pull request by CrashAndSideburns**
## Description

Some changes to the grammar of the language and appropriate parts of the lexer to allow for numeric literals to have leading zeros. Previously, a numeric literal such as `001` would have been disallowed due to conflict with the syntax for octal literals, but this is now allowable due to a change made to the syntax for octal literals.

To effect this change, whenever a numeric literal is lexed which begins with a `0` after any relevant prefix (such as `0x`), these zeros are dropped before the literal is passed on to the appropriate lexing function. This also allows for the `TT_LIT_INT_0` token to be removed, since the lexing of zeros is now merged with the lexing of all other decimal literals.

## Motivation and Context

This change allows for leading zeros in literals of all types, which is frequently desirable for aligning numeric literals in source code for legibility reasons.

## How Has This Been Tested

All current tests pass.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] My code follows the code style of this project.
- [x] I have run clang-format to make sure that it does.
- [x] My code passes all of the tests.
- [x] I have added tests to cover any new features added.
- [x] I have updated the documentation to reflect the current state of the repo.
- [x] I have removed or updated comments that are no longer relevant, wherever I saw them.
- [x] This included updating the copyright header.
- [x] I am aware that I am releasing my code under the GPL Version 3 or later.
- [x] I agree to the Developer Certificate of Origin (<https://developercertificate.org/>).

### CODE REVIEW

File: src/main/lexer/lexer.c [hunk 1]
```diff
@@ -559,38 +559,63 @@ static void lexNumber(FileListEntry *entry, Token *token) {
...
+    // maybe decimal, hex, dec (float), oct, bin, or zero
     char second = get(state);
     switch (second) {
       case 'b': {
         // [+-]0b
-        // is binary, lex it
+        // is binary, remove leading zeros and lex it
+        while ((next = get(state)) == '0')
+          ;
+        put(state, 1);
```
Comment (JustinHuPrime) [inline src/main/lexer/lexer.c:570]: I don't think this is needed

File: src/main/lexer/lexer.c [hunk 2]
```diff
@@ -559,38 +559,63 @@ static void lexNumber(FileListEntry *entry, Token *token) {
...
+          lexDecimal(entry, token, start);
+          return;
         } else {
-          // just a zero - [+-]0
+          // just a zero - [+-]0+
           put(state, 1);
-          clip(state, token, start, TT_LIT_INT_0);
+          clip(state, token, start, TT_LIT_INT_D);
           return;
         }
```
Comment (JustinHuPrime) [inline src/main/lexer/lexer.c:620]: These three cases probably collapse into one or maybe two cases

File: src/test/tests/lexer.c [hunk 3]
```diff
@@ -156,7 +156,7 @@ static void testAllTokens(void) {
       {TT_LIT_INT_B, 10, 71, "0b1"},
       {TT_LIT_INT_O, 10, 74, "+0o377"},
 
-      {TT_LIT_INT_0, 14, 1, "0"},
+      {TT_LIT_INT_D, 14, 1, "0"},
```
Comment (JustinHuPrime) [inline src/test/tests/lexer.c:159]: Your probably want to add tests for leading zero variants of the literals

File: standard/Appendix A - Lexicon.ebnf [hunk 4]
```diff
@@ -47,11 +47,10 @@ magic_token = "__FILE__"
             | "__VERSION__"
 ;
 
-int_literal = [ "-" | "+" ], ( nonzero_digit, { digit }
+int_literal = [ "-" | "+" ], ( { digit }
```
Comment (JustinHuPrime) [inline standard/Appendix A - Lexicon.ebnf:50]:
```suggestion
int_literal = [ "-" | "+" ], ( digit, { digit }
```
Otherwise you allow the empty string to be a int literal

**Review by JustinHuPrime — CHANGES_REQUESTED** [review]
